client/state/http/index.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import superagent from 'superagent';
+
+/***
+ * Internal dependencies
+ */
+import { extendAction } from 'state/utils';
+import { HTTP_REQUEST } from 'state/action-types';
+import { failureMeta, successMeta } from 'state/data-layer/wpcom-http';
+
+const isAllHeadersValid = headers =>
+	headers.every( headerPair => Array.isArray( headerPair ) &&
+		headerPair.length === 2 &&
+		typeof headerPair[ 0 ] === 'string' &&
+		typeof headerPair[ 1 ] === 'string'
+	);
+
+/***
+ * Handler to perform an http request based on `HTTP_REQUEST` action parameters:
+ * {String} url the url to request
+ * {String} method the method we should use in the request: GET, POST etc.
+ * {Array<Array<String>>} headers array of [ 'key', 'value' ] pairs for the request headers
+ * {Object} body data send as body
+ * {Boolean} withCredentials save cookie set on request
+ * {Action} onSuccess action to dispatch on success with data meta
+ * {Action} onFailure action to dispatch on failure with error meta
+ *
+ * @param {Function} dispatch redux store dispatch
+ * @param {Object} action dispatched action we need to handle
+ * @returns {Promise} promise of the handled request
+ */
+const httpHandler = ( { dispatch }, action ) => {
+	const {
+		url,
+		method,
+		headers = [],
+		body,
+		withCredentials,
+		onSuccess,
+		onFailure
+	} = action;
+
+	if ( ! isAllHeadersValid( headers ) ) {
+		const error = new Error( "Not all headers were of an array pair: [ 'key', 'value' ]" );
+		dispatch( extendAction( onFailure, failureMeta( error ) ) );
+		return Promise.reject( error );
```

Comment:
I think we spoke about this before, but is there any way we could get away with _not_ returning the promise here? we can already test this code just fine by creating a dispatch spy, so I don't think it aids with testing. When we provide a meaningful return value then I think it leads people to _use_ the return value and then _rely_ on the return value and that leads us into code patterns which break the way the data layer is designed to work. For example, suppose we had a request in-flight and decided to drop this one: some UI component depending on the return value will probably sit there waiting for a request to return that never does. When queuing while offline as well this same problem would appear.